### ISSUE TITLE
[Request]: Save a permanent local copy of listening history

### ISSUE THREAD

**Issue by UnviableFriend** [lyarenei/jellyfin-plugin-listenbrainz#88]
### Description

Save a permanent local copy of listens (Optionally?).

### Proposed solution

Even a JSON file per user would work well, I think. My listens are largely imported from last.fm, so there isn't as many details in them, but my 350k+ downloaded listen data from LB only uses 304MB. So space shouldn't really be a big concern.

### Additional context

There was a small period where LB was offline the day before this issue was made, and in the past there has been issues that have resulted in a loss of some listens. Things like this are likely to happen despite how hard the devs work to keep everything working smoothly.
Additionally a person may lose their account info, or delete their account, or listen history, for one reason or another and want a way to recover it later.
A local realtime backup of listens would give peace of mind and simplify recovery of data if the need arises.

**Comment by lyarenei** [comment]
> There was a small period where LB was offline the day before this issue was made, and in the past there has been issues that have resulted in a loss of some listens. Things like this are likely to happen despite how hard the devs work to keep everything working smoothly.

This is exactly what the [listen cache feature](https://github.com/lyarenei/jellyfin-plugin-listenbrainz/blob/main/doc/how-it-works.md#listen-cache) is for - if the listen submissions fail, then they are saved and resubmitted later. Unfortunately, this was broken among other things in the version `3.4.0.0` (#87) so I get where are you coming from.

> Additionally a person may lose their account info, or delete their account, or listen history, for one reason or another and want a way to recover it later.

I believe this is more on the user than anything. All of these situations are ultimately the reason why one should regularly back up their data, especially if it holds some significance for them. The user can always [export their listens and feedbacks](https://listenbrainz.org/settings/export/).

> A local realtime backup of listens would give peace of mind and simplify recovery of data if the need arises.

Although this is a bit nitpicky, you could argue that you can lose the access to the server as well (unless you are not running Jellyfin locally on your network). But other than that, I think some sort of tracking has been probably already done by some external monitoring tools. Back when I used Plex, I also used Tautulli and it tracked everything, provided graphs and so on, it was really nice. I'm not sure, but I wouldn't be surprised if someone already did similar tool for Jellyfin.

On the other hand, if you wanted to keep it simple, then probably a Jellyfin plugin entirely dedicated to that would be the way to go. Maybe the [playback reporting plugin](https://github.com/jellyfin/jellyfin-plugin-playbackreporting) could fit the use-case?

With the plugin approach, there is also an issue with data storage location where you would usually save the json file. Basically when you upgrade plugin, the plugin data folder changes (the plugin version is a part of the folder name) and so you would have your activity history all over place. Of course this can be worked around by just having the file elsewhere, but that's not a good practice. You also need to track the file and make sure it's deleted when the plugin is uninstalled where with the plugin data directory, the whole folder just gets deleted. You could send data to somewhere else, but that just increases complexity.

---

TL;DR, unless I am missing something, this is something where a separate plugin would make more sense. Especially as natural next step for such plugin would be to track all activity, not just music playback. Assuming there isn't an already existing solution which would do the job.

Let me know what you think.

**Comment by lyarenei** [comment]
> There was "[the incident](https://blog.metabrainz.org/2023/11/04/listenbrainz-service-incident/)" as an example of times where this would be especially helpful.

Ah, I see. I didn't even know about that tbh. But yeah, if ListenBrainz has to do a DB restore, then the user backup is the only option to get back the data.

> A local copy in the same format as it's being submitted would be very easy to grab data from any time frame that was lost on servers and resubmit.
That could also be helped by users backing up their data, but I doubt most know they can, let alone would remember to do it regularly(Figuring out how to automate mine is on my todo list).

> That said, I guess it could be a better feature request to have playback reporting implement export features(I think currently it can only do a full backup of everything?)

Hmm, I think I might have misunderstood what you meant at first. I thought that you meant some sort of generic playback history keeping (and I was focused on the generic part a bit too much), which is something that does not fit into this plugin very well - and that's why I suggested the playback reporting plugin.

But if reword this into let's say `automatic ListenBrainz data backup`, then it sounds much nicer (to me). 

---

Some implementation notes:
- Backup frequency
  - Each user may want to have a different frequency, so some sort of a system for dynamically adding and removing scheduled tasks for each user may be a way. But that might not be possible within the plugin API limits or Jellyfin as a whole. It also sounds unnecessarily complicated.   
  - Just sync all users (like with favorites at this moment) and be done with it.
- Backup files
  - Offer to select the path, just like in the playback reporting plugin.
    - If path is not set, feature is disabled (disk space in `$CONFIG` may be limited)
  - Cover cases when the backup is not successful, so the previous backup is not lost.
  - Splitt into multiple files (working with large files is suboptimal at best).
- Backup data format
  - JSON - as received from ListenBrainz. 
    - Maybe throw away some fields which are not critical and can be fetched again if necessary. 
    - Minification can help with the file size as well.
- Import feature
  - Provided by ListenBrainz => no implementation, governs saved data from exports
  - Provided by plugin => needs implementation, free to choose export formats

**Comment by lyarenei** [comment]
@UnviableFriend It has been quite some time, sorry. If you are still interested in this, you can switch to the dev repo and try plugin version `4.1.0.5`, where I implemented the backup functionality in the way you originally suggested. Originally, I thought downloading backups was the approach to go with, but then I realized there are some situations which cannot be covered by this approach.

Please let me know what do you think if you decide to test it. Thank you.

---
To address notes from https://github.com/lyarenei/jellyfin-plugin-listenbrainz/issues/88#issuecomment-1967678187

> Backup frequency

- Not necessary. Backups are real-time.
- Admins can select users which should be backed up.

> Backup files

- If path is not set, the feature cannot be enabled.
- Saved listens are split into files per user per day.

> Backup data format

- JSON
- Pure listen data (no Jellyfin metadata), ready for manual submission to ListenBrainz.

> Import feature

- Not implemented on ListenBrainz to this date (use submission endpoint with `import` type). 
- Not implemented in plugin, can be easily worked around with a simple script.

**Comment by Maxr1998** [comment]
> > Backup data format
> 
> - JSON
> - Pure listen data (no Jellyfin metadata), ready for manual submission to ListenBrainz.

Does this listen data include additional metadata queried from MusicBrainz, or is it really just the raw data from Jellyfin? Would it be an option to include the MusicBrainz data if it's possible to connect to MB, with only falling back to the "pure" data.

**Comment by lyarenei** [comment]
> Does this listen data include additional metadata queried from MusicBrainz, or is it really just the raw data from Jellyfin? Would it be an option to include the MusicBrainz data if it's possible to connect to MB, with only falling back to the "pure" data.

Yes, all metadata available at the moment of saving are saved - the same what is also sent to ListenBrainz shortly after.

**Comment by Maxr1998** [comment]
That's awesome. I'll give it a go in the next few days.
